File: tba/src/main/java/de/gaalop/productcomputer2/InnerProductCalculator.java
package de.gaalop.productcomputer2;

import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;

/**
 * Implements the ProductCalculator interface for the inner product
 * @author christian
 */
public class InnerProductCalculator implements ProductCalculator {

    private HashMap<Integer, Integer> baseSquares;

    public SumOfBlades calculate(Blade blade1, Blade blade2, HashMap<Integer, Integer> baseSquares) {
        this.baseSquares = baseSquares;
        
        LinkedList<Integer> base1 = blade1.getBaseVectors();
        LinkedList<Integer> base2 = blade2.getBaseVectors();

        if (base1.size() == 0 || base2.size() == 0)
            return new SumOfBlades(new LinkedList<Blade>());
        


        Blade result;
        if (base1.size() == 1) {
            if (base2.size() == 1)
                result = calculateProductBlades(base1.getFirst(), base2.getFirst());
            else
                result = calculateProductBlades(base1.getFirst(), base2);
        } else {
            if (base2.size() == 1)
                result = calculateProductBlades(base1, base2.getFirst());
            else
                result = calculateProductBlades(base1, base2);
        }
        result.setPrefactor(result.getPrefactor()*blade1.getPrefactor()*blade2.getPrefactor());

        return new SumOfBlades(result);

    }

    /**
     * Calculates a product of two bases, which have both more than one base element
     * @param base1 The first base
     * @param base2 The second base
     */
    private Blade calculateProductBlades(LinkedList<Integer> base1, LinkedList<Integer> base2) {


        if (base1.size() > base2.size()) {
            // Bl*ak, k<l

            Iterator<Integer> it = base2.listIterator();
            Blade result = calculateProductBlades(base1, it.next());
            while (it.hasNext()) {
                Blade result2 = calculateProductBlades(result.getBaseVectors(), it.next());
                result2.setPrefactor(result2.getPrefactor()*result.getPrefactor());
                result = result2;
            }

            return result;
        } else {
            // ak*bl, k<l
            Iterator<Integer> it = base1.descendingIterator();
            Blade result = calculateProductBlades(it.next(), base2);

            while (it.hasNext()) {
                Blade result2 = calculateProductBlades(it.next(), result.getBaseVectors());
                result2.setPrefactor(result2.getPrefactor()*result.getPrefactor());
                result = result2;
            }


            return result;
        }

    }

    /**
     * Returns the index of an int in an int arr
     * @param search The searched int
     * @param arr The array to search in
     * @return The index, -1 if the array doesn't contain the searched int
     */
    private static int getIndexOfString(Integer search, LinkedList<Integer> arr) {
        int i=0;
        for (Integer j: arr) {
            if (search == j)
                return i;
            i++;
        }
        return -1;
    }

    /**
     * Calculates a product of a base element and a base
     * @param base1 The base element
     * @param base2 The base
     * @return The inner product as Expression
     */
    private Blade calculateProductBlades(Integer base1, LinkedList<Integer> base2) {
        int index = getIndexOfString(base1, base2);
        if (index > -1) {
            LinkedList<Integer> arr = new LinkedList<Integer>();
            int j=0;
            for (Integer i: base2) {
                if (j != index)
                    arr.add(i);
                j++;
            }

            float prefactor = 1;
            if ((index%2) == 1)
                prefactor *= -1;

            if (baseSquares.get(base1) == -1)
                prefactor *= -1;

            return new Blade(prefactor, arr);
        } else
            return new Blade(0, new LinkedList<Integer>());
    }

    /**
     * Calculates a product of a base and a base element
     * @param base1 The base
     * @param base2 The base element
     * @return The inner product as Expression
     */
    private Blade calculateProductBlades(LinkedList<Integer> base1, Integer base2) {
        int index = getIndexOfString(base2, base1);
        if (index > -1) {
            LinkedList<Integer> arr = new LinkedList<Integer>();
            int j=0;
            for (Integer i: base1) {
                if (j != index)
                    arr.add(i);
                j++;
            }

            float prefactor = 1;
            if (((base1.size()-index-1) %2) == 1)
                prefactor *= -1;

            if (baseSquares.get(base2) == -1)
                prefactor *= -1;

            return new Blade(prefactor, arr);
        } else
            return new Blade(0, new LinkedList<Integer>());
    }

    /**
     * Calculates a product two base elements
     * @param base1 The base element
     * @param base2 The base element
     * @return The inner product as Expression
     */
    private Blade calculateProductBlades(int base1, int base2) {
        if (base1 == base2)
            return new Blade((base1 != 0) ? baseSquares.get(base1) : 0, new LinkedList<Integer>());
        return new Blade(0, new LinkedList<Integer>());
    }


}

File: productComputerC/InnerProductCalculator.cpp
/*
 * InnerProductCalculator.cpp
 *
 *  Created on: 22.11.2011
 *      Author: christian
 */

#include "InnerProductCalculator.h"

InnerProductCalculator::InnerProductCalculator() {
}

InnerProductCalculator::~InnerProductCalculator() {
}

void InnerProductCalculator::calculate(const Blade& blade1,const  Blade& blade2, unordered_map<int,int>& baseSquares, SumOfBlades& result) {
        this->baseSquares = baseSquares;

        const intVector& base1 = blade1.baseVectors;
        const intVector& base2 = blade2.baseVectors;

        if (base1.size() == 0 || base2.size() == 0)
        	return;

        if (base1.size() == 1) {
            if (base2.size() == 1) {
            	Blade resultB;
                calculateProductBlades(base1.front(), base2.front(), resultB);
                resultB.setPrefactor(resultB.getPrefactor()*blade1.getPrefactor()*blade2.getPrefactor());
                result.addBlade(resultB);
                return;
            }
            else {
                Blade resultB;
                calculateProductBlades(base1.front(), base2, resultB);
                resultB.setPrefactor(resultB.getPrefactor()*blade1.getPrefactor()*blade2.getPrefactor());
                result.addBlade(resultB);
                return;
				}
        } else {
            if (base2.size() == 1) {
                Blade resultB;
                calculateProductBlades(base1, base2.front(),resultB);
                resultB.setPrefactor(resultB.getPrefactor()*blade1.getPrefactor()*blade2.getPrefactor());
                result.addBlade(resultB);
                return;
				}
            else {
                Blade resultB;
                calculateProductBlades(base1, base2,resultB);
                resultB.setPrefactor(resultB.getPrefactor()*blade1.getPrefactor()*blade2.getPrefactor());
                result.addBlade(resultB);
                return;
				}
        }
    }

    /**
     * Calculates a product of two bases, which have both more than one base element
     * @param base1 The first base
     * @param base2 The second base
     */
    void InnerProductCalculator::calculateProductBlades(const intVector& base1, const intVector& base2, Blade& resultF) {


        if (base1.size() > base2.size()) {
            // Bl*ak, k<l

            intVector::const_iterator it = base2.begin();
            Blade result;
            calculateProductBlades(base1, *it, result);
            resultF = result;
            ++it;
            if (it == base2.end()) return;
            do {

            	intVector& baseVectors = result.baseVectors;
                Blade result2;
                calculateProductBlades(baseVectors, *it, result2);
                result2.setPrefactor(result2.getPrefactor()*result.getPrefactor());
                resultF = result2;
                result = result2;
                ++it;
            } while (it != base2.end());

        } else {
            // ak*bl, k<l
        	intVector::const_reverse_iterator it = base1.rbegin();
            Blade result;
            calculateProductBlades(*it, base2, result);

            ++it;
            resultF = result;
            if (it == base1.rend()) return;
            do {
            	intVector& baseVectors = result.baseVectors;
                Blade result2;
                calculateProductBlades(*it, baseVectors, result2);
                result2.setPrefactor(result2.getPrefactor()*result.getPrefactor());
                resultF = result2;
                result = result2;
                ++it;
            } while (it != base1.rend());
        }

    }



    /**
     * Calculates a product of a base element and a base
     * @param base1 The base element
     * @param base2 The base
     * @return The inner product as Expression
     */
    void InnerProductCalculator::calculateProductBlades(const int base1, const intVector& base2, Blade& result) {
        int index = VectorMethods::getIndexOfString(base1, base2);
        if (index > -1) {
            intVector& arr = result.baseVectors;
            arr.clear();
            int j=0;
            for (intVector::const_iterator i=base2.begin(); i != base2.end(); ++i) {
                if (j != index)
                    arr.push_back(*i);
                j++;
            }

            float prefactor = 1;
            if ((index%2) == 1)
                prefactor *= -1;

            if (baseSquares[base1] == -1)
                prefactor *= -1;

            result.setPrefactor(prefactor);
        } else {
        	result.baseVectors.clear();
        	result.setPrefactor(0);
        }
    }

    /**
     * Calculates a product of a base and a base element
     * @param base1 The base
     * @param base2 The base element
     * @return The inner product as Expression
     */
    void InnerProductCalculator::calculateProductBlades(const intVector& base1, const int base2, Blade& result) {
        int index = VectorMethods::getIndexOfString(base2, base1);
        if (index > -1) {
        	intVector& arr = result.baseVectors;
        	arr.clear();
            int j=0;
            for (intVector::const_iterator i=base1.begin(); i != base1.end(); ++i) {
                if (j != index)
                    arr.push_back(*i);
                j++;
            }

            float prefactor = 1;
            if (((base1.size()-index-1) %2) == 1)
                prefactor *= -1;

            if (baseSquares[base2] == -1)
                prefactor *= -1;

            result.setPrefactor(prefactor);
        } else {
        	result.baseVectors.clear();
            result.setPrefactor(0);
        }
    }

    /**
     * Calculates a product two base elements
     * @param base1 The base element
     * @param base2 The base element
     * @return The inner product as Expression
     */
    void InnerProductCalculator::calculateProductBlades(const int base1, const int base2, Blade& result) {
    	result.baseVectors.clear();
        if (base1 == base2 && base1 != 0) {
			result.setPrefactor(baseSquares[base1]);
			return;
        }
		result.setPrefactor(0);
    }
